### PR TITLE
Warn that implicit should have explicit type [ci: last-only]

### LINKED ISSUE #5265: warn on implicit def without explicit result type
The following code snipet : https://gist.github.com/1427587
does not compile.

```scala
import java.util.Date

trait TDate 

trait TT[A1,T1]

trait TTFactory[F,G] {
  def create(f: F) : TT[F,G]
  def sample: F
}

object Impls {

  // If the c1 is declared before c2, it compiles fine
  // or if the implicit's result type is specified explicitly
  implicit def c2(s: Date)/* : TT[Date, TDate] */ = c1.create(s)  

  implicit val c1 = new TTFactory[Date,TDate] {
    def create(v: Date): TT[Date,TDate] = sys.error("")
    def sample = new Date
  }
}
```

### PR DESCRIPTION
**Pull request by som-snytt**
Implicit definitions are required to have an explicit type ascribed in Scala 3. Now Scala 2 warns if the type is missing (and therefore inferred). Under `-Xsource:3`, the warning becomes an error.

Local implicits are exempt because they are not available before they are defined.

The warnings may be suppressed globally with `-Wconf:cat=other-implicit-type:s` or locally using `@annotation.nowarn("cat=other-implicit-type")`.

Note that a whitebox macro may be declared of type `Any` and arbitrarily narrowed by its expansion.

Fixes scala/bug#5265

### CODE REVIEW

**Comment by som-snytt** [comment]
`-Werror` FTW.
```
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/src/reflect/scala/reflect/api/Internals.scala:1101:18: Implicit definition should have explicit type
[error]     implicit val token = new CompatToken
[error]                  ^
[error] one error found
```

File: test/files/run/t5565.scala [hunk 1]
```diff
@@ -1,7 +1,9 @@
+import scala.annotation.nowarn
 import scala.language.reflectiveCalls
 import scala.language.implicitConversions
 
 object Test extends App {
+  @nowarn  // the inferred type includes the default arg, which can't be written explicitly
```
Comment (som-snytt) [inline test/files/run/t5565.scala:6]: @dwijnand inviting your thoughts on your example: implicits should have explicit types, but it is willing to infer a type that can't be written explicitly. In this case, the default arg is not allowed in the refinement. (One could introduce a trait, but that would negate the test.) Probably you will say: "Rules are made to be broken."

**Comment by som-snytt** [comment]
Fields are not marked implicit, and can't access the accessor without cycles, so mark it and check when typing the getter.

**Comment by SethTisue** [comment]
/rebuild

(https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/16292 got stuck)

meanwhile, Travis-CI has failure information

**Comment by som-snytt** [comment]
I didn't switch to draft after handling `val`, and still doesn't handle `val` in traits. Plus cycle of test breakage.

File: test/files/run/t5565.scala (hunk 1 above)
Comment (dwijnand) [reply test/files/run/t5565.scala:6]: Nice find!

**Comment by som-snytt** [comment]
My favorite conflicts are `JavaUniverseForce` and `StdAttachments`.

**Comment by som-snytt** [comment]
Hung again. I don't see the red ex for canceling jenkins job.

Travis:
```
[info] + IndexScript.allPackages: OK, proved property.
[info] Elapsed time: 0.432 sec 
failing seed for mutable.TreeMapProjection.get, contains is WGrJJZh_6eL-dRMg_fX5Komy1UboU3eTd2SlGrJ3d9G=
[info] ! mutable.TreeMapProjection.get, contains: Exception raised on property evaluation.
[info] > ARG_0: HashMap( -> 0)
[info] > ARG_1: None
[info] > ARG_2: Some()
[info] > ARG_0_ORIGINAL: HashMap(պ -> -2147483648)
[info] > ARG_1_ORIGINAL: Some()
[info] > Exception: java.lang.NullPointerException: null
[info] scala.collection.mutable.TreeMap$TreeMapProjection.isInsideViewBounds(TreeMap.scala:193)
[info] scala.collection.mutable.TreeMap$TreeMapProjection.contains(TreeMap.scala:211)
[info] scala.collection.mutable.MutableTreeMapProjectionProperties$.$anonfun$new$125(MutableTreeMap.scala:238)
[info] scala.collection.mutable.MutableTreeMapProjectionProperties$.$anonfun$new$125$adapted(MutableTreeMap.scala:237)
[info] scala.collection.IterableOnceOps.forall(IterableOnce.scala:589)
[info] scala.collection.IterableOnceOps.forall$(IterableOnce.scala:586)
[info] scala.collection.AbstractIterable.forall(Iterable.scala:933)
[info] scala.collection.mutable.MutableTreeMapProjectionProperties$.$anonfun$new$124(MutableTreeMap.scala:237)
[info] scala.collection.mutable.MutableTreeMapProjectionProperties$.$anonfun$new$124$adapted(MutableTreeMap.scala:230)
```

Locally,
```
[info] + Range inclusiveByOne.foreach.visited.size: OK, passed 100 tests.
[info] Elapsed time: 0.002 sec
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fad5f9c03fc, pid=30176, tid=30206
#
# JRE version: OpenJDK Runtime Environment (18.0.1.1+2) (build 18.0.1.1+2-6)
# Java VM: OpenJDK 64-Bit Server VM (18.0.1.1+2-6, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0xd263fc][info] + Range inclusiveByOne.sum: OK, passed 100 tests.
[info] Elapsed time: 0.002 sec
[info] + Range inclusiveByOne.sumCustomNumeric: OK, passed 100 tests.
[info] Elapsed time: 0.004 sec
```
Travis says both
```
export ADOPTOPENJDK=8
$ java -Xmx32m -version
openjdk version "11.0.2" 2019-01-15
OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)
$ javac -J-Xmx32m -version
javac 11.0.2
```

**Comment by som-snytt** [comment]
I should have @SethTisue about cancel culture.

**Comment by SethTisue** [comment]
canceled another stuck Jenkins job just now.
signed, The Canceler

**Comment by som-snytt** [comment]
Oops sorry @SethTisue I forgot to diagnose the hang slash crash. I pushed so I could view diffs.

While I was mechanically adding inferred types, I didn't notice:

```
-  implicit val ord = implicitly[Ordering[K]]
+  private val ord = implicitly[Ordering[K]]
```

**Comment by som-snytt** [comment]
A clean repo crashes in `scalacheck/test` on latest 18. It would be nice if you could tell which test.
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f87d3cd4acc, pid=1203, tid=1219
#
# JRE version: OpenJDK Runtime Environment (18.0.2.1+1) (build 18.0.2.1+1-1)
# Java VM: OpenJDK 64-Bit Server VM (18.0.2.1+1-1, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0xd27acc]  SubTypeCheckNode::sub(Type const*, Type const*) const+0x26c

Current thread (0x00007f87cc0d0830):  JavaThread "C2 CompilerThread0" daemon [_thread_in_native, id=1219, stack(0x00007f87a84ad000,0x00007f87a85ae000)]


Current CompileTask:
C2:   5155 9279       4       scala.collection.ArrayOps$::prepended$extension (48 bytes)
```

**Comment by som-snytt** [comment]
It's the old space after `[ci:]` gotcha!

I'll re-squash after @lrytz review.

**Comment by SethTisue** [comment]
> A clean repo crashes in scalacheck/test on latest 18

mind opening a scala-dev ticket on that?

**Review by lrytz — COMMENTED** [review]
The diff LGTM, and it's great to have this warning, thank you!

I am a bit worried about enabling it by default. Maybe if we introduce a new warning category and publish the opt-out flag in the release notes, it's OK? As far as I can tell, the opt out right now would be `"-Wconf:msg=Implicit definition should have explicit type:s"`.

**Comment by som-snytt** [comment]
Could add a lint option if not `-Xsource:3`? Usually it's deprecation then error.

**Comment by SethTisue** [comment]
let's figure this out for ~2.13.10~ 2.13.11

**Comment by SethTisue** [comment]
@som-snytt mind resolving the merge conflict? where does this stand, do you believe it is ready for review?

**Comment by SethTisue** [comment]
@som-snytt /me turns on the machine that goes "ping!"

**Review by lrytz — APPROVED** [review]
LGTM, thank you!

@SethTisue what's your view on making the warning on by default with `-Wconf` / `@nowarn` opt-out?

**Review by SethTisue — APPROVED** [review]
> Maybe if we introduce a new warning category and publish the opt-out flag in the release notes, it's OK? 

Yes, I think that's sufficient.

**Comment by SethTisue** [comment]
Note that this impacts macro authors too; see https://github.com/scala/bug/issues/12798

**Pull request merged**